File: functions/groups/get_group_feed.py
from firebase_admin import firestore
from flask import Request, abort
from models.constants import Collections, GroupFields, QueryOperators, UpdateFields
from models.data_models import FeedResponse, Update
from utils.logging_utils import get_logger


def get_group_feed(request: Request, group_id: str) -> FeedResponse:
    """
    Retrieves all updates for a specific group, paginated.
    
    This function fetches updates that include the specified group ID in their group_ids array.
    The updates are returned in descending order by creation time (newest first) and
    support pagination for efficient data loading.
    
    Args:
        request: The Flask request object containing:
                - user_id: The authenticated user's ID (attached by authentication middleware)
                - validated_params: Pagination parameters containing:
                    - limit: Maximum number of updates to return
                    - after_timestamp: Timestamp for pagination
        group_id: The ID of the group to retrieve updates for
    
    Query Parameters:
        - limit: Maximum number of updates to return (default: 20, min: 1, max: 100)
        - after_timestamp: Timestamp for pagination in ISO format (e.g. "2025-01-01T12:00:00Z")
    
    Returns:
        A FeedResponse containing:
        - A list of updates for the specified group
        - A next_timestamp for pagination (if more results are available)
        
    Raises:
        404: Group not found
        403: User is not a member of the group
        500: Internal server error
    """
    logger = get_logger(__name__)
    logger.info(f"Retrieving feed for group: {group_id}")

    # Get the authenticated user ID from the request
    current_user_id = request.user_id

    # Initialize Firestore client
    db = firestore.client()

    # Get pagination parameters from the validated request
    validated_params = request.validated_params
    limit = validated_params.limit if validated_params else 20
    after_timestamp = validated_params.after_timestamp if validated_params else None

    logger.info(f"Pagination parameters - limit: {limit}, after_timestamp: {after_timestamp}")

    # First, check if the group exists and if the user is a member
    group_ref = db.collection(Collections.GROUPS).document(group_id)
    group_doc = group_ref.get()

    if not group_doc.exists:
        logger.warning(f"Group {group_id} not found")
        abort(404, description="Group not found")

    group_data = group_doc.to_dict()
    members = group_data.get(GroupFields.MEMBERS, [])

    # Check if the current user is a member of the group
    if current_user_id not in members:
        logger.warning(f"User {current_user_id} is not a member of group {group_id}")
        abort(403, description="You must be a member of the group to view its feed")

    # Build the query for updates from this group
    query = db.collection(Collections.UPDATES) \
        .where(UpdateFields.GROUP_IDS, QueryOperators.ARRAY_CONTAINS, group_id) \
        .order_by(UpdateFields.CREATED_AT, direction=firestore.Query.DESCENDING)

    # Apply pagination if an after_timestamp is provided
    if after_timestamp:
        query = query.start_after({UpdateFields.CREATED_AT: after_timestamp})
        logger.info(f"Applying pagination with timestamp: {after_timestamp}")

    # Apply limit last
    query = query.limit(limit)

    # Execute the query
    docs = query.stream()
    logger.info("Query executed successfully")

    updates = []
    last_timestamp = None

    # Process the query results
    for doc in docs:
        doc_data = doc.to_dict()
        created_at = doc_data.get(UpdateFields.CREATED_AT, "")

        # Track the last timestamp for pagination
        if created_at:
            last_timestamp = created_at

        # Convert Firestore document to Update model
        updates.append(Update(
            updateId=doc.id,
            created_by=doc_data.get(UpdateFields.CREATED_BY, ""),
            content=doc_data.get(UpdateFields.CONTENT, ""),
            group_ids=doc_data.get(UpdateFields.GROUP_IDS, []),
            sentiment=doc_data.get(UpdateFields.SENTIMENT, 0),
            created_at=created_at
        ))

    # Set up pagination for the next request
    next_timestamp = None
    if last_timestamp and len(updates) == limit:
        next_timestamp = last_timestamp
        logger.info(f"More results available, next_timestamp: {next_timestamp}")

    logger.info(f"Retrieved {len(updates)} updates for group: {group_id}")
    return FeedResponse(
        updates=updates,
        next_timestamp=next_timestamp
    )

File: functions/src/groups/get-group-feed.ts
import { Request, Response } from 'express';
import { getFirestore, QueryDocumentSnapshot } from 'firebase-admin/firestore';
import { Collections, QueryOperators } from '../models/constants.js';
import { FeedResponse, PaginationPayload } from '../models/data-models.js';
import { FeedDoc, groupConverter, UpdateDoc, updateConverter, uf } from '../models/firestore/index.js';
import { ForbiddenError, NotFoundError } from '../utils/errors.js';
import { getLogger } from '../utils/logging-utils.js';
import { applyPagination, generateNextCursor, processQueryStream } from '../utils/pagination-utils.js';
import { fetchUpdatesByIds, processEnrichedFeedItems } from '../utils/update-utils.js';

import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const logger = getLogger(path.basename(__filename));

/**
 * Retrieves all updates for a specific group, paginated.
 *
 * This function fetches updates that include the specified group ID in their group_ids array.
 * The updates are returned in descending order by creation time (newest first) and
 * support pagination for efficient data loading.
 *
 * @param req - The Express request object containing:
 *              - userId: The authenticated user's ID (attached by authentication middleware)
 *              - validated_params: Pagination parameters containing:
 *                - limit: Maximum number of updates to return
 *                - after_cursor: Cursor for pagination (base64 encoded document path)
 * @param res - The Express response object
 * @param groupId - The ID of the group to retrieve updates for
 *
 * Query Parameters:
 * - limit: Maximum number of updates to return (default: 20, min: 1, max: 100)
 * - after_cursor: Cursor for pagination (base64 encoded document path)
 *
 * @returns A FeedResponse containing:
 * - A list of updates for the specified group
 * - A next_cursor for pagination (if more results are available)
 *
 * @throws 404: Group not found
 * @throws 403: User is not a member of the group
 * @throws 500: Internal server error
 */
export const getGroupFeed = async (req: Request, res: Response, groupId: string): Promise<void> => {
  logger.info(`Retrieving feed for group: ${groupId}`);

  // Get the authenticated user ID from the request
  const currentUserId = req.userId;

  // Initialize Firestore client
  const db = getFirestore();

  // Get pagination parameters from the validated request
  const validatedParams = req.validated_params as PaginationPayload;
  const limit = validatedParams?.limit || 20;
  const afterCursor = validatedParams?.after_cursor;

  logger.info(`Pagination parameters - limit: ${limit}, after_cursor: ${afterCursor}`);

  // First, check if the group exists and if the user is a member
  const groups = db.collection(Collections.GROUPS).withConverter(groupConverter);
  const groupRef = groups.doc(groupId);
  const groupDoc = await groupRef.get();

  const groupData = groupDoc.data();
  if (!groupData) {
    logger.warn(`Group ${groupId} not found`);
    throw new NotFoundError('Group not found');
  }

  const members = groupData.members || [];

  // Check if the current user is a member of the group
  if (!members.includes(currentUserId)) {
    logger.warn(`User ${currentUserId} is not a member of group ${groupId}`);
    throw new ForbiddenError('You must be a member of the group to view its feed');
  }

  // Build the query for updates from this group
  let query = db
    .collection(Collections.UPDATES)
    .withConverter(updateConverter)
    .where(uf('group_ids'), QueryOperators.ARRAY_CONTAINS, groupId)
    .orderBy(uf('created_at'), QueryOperators.DESC);

  // Apply cursor-based pagination - Express will automatically catch errors
  const paginatedQuery = await applyPagination(query, afterCursor, limit);

  // Process updates using streaming
  const { items: updateDocs, lastDoc } = await processQueryStream<QueryDocumentSnapshot>(
    paginatedQuery,
    (doc) => doc,
    limit,
  );

  // Get all update IDs from the query results
  const updateIds = updateDocs.map((updateDoc) => updateDoc.id);

  // Fetch update data using util
  const updateMap = await fetchUpdatesByIds(updateIds);

  // Convert update documents to feed documents for processEnrichedFeedItems
  const feedDocs = updateDocs.map((updateDoc) => {
    const updateData = updateDoc.data() as UpdateDoc;
    const feedData: FeedDoc = {
      update_id: updateDoc.id,
      created_at: updateData.created_at,
      created_by: updateData.created_by,
      direct_visible: false,
      friend_id: undefined,
      group_ids: updateData.group_ids || [],
    };
    return {
      id: updateDoc.id,
      data: () => feedData,
    } as QueryDocumentSnapshot<FeedDoc>;
  });

  // Use util to enrich updates
  const enrichedUpdates = await processEnrichedFeedItems(feedDocs, updateMap);

  logger.info('Query executed successfully');

  // Set up pagination for the next request
  const nextCursor = generateNextCursor(lastDoc, enrichedUpdates.length, limit);

  logger.info(`Retrieved ${enrichedUpdates.length} updates for group: ${groupId}`);

  // Return the response
  const response: FeedResponse = {
    updates: enrichedUpdates,
    next_cursor: nextCursor,
  };

  res.json(response);
};
